src/Host/Client/Impl/BrokerServices/WebService.cs

```diff
@@ -15,71 +15,123 @@
+                        if (Enum.TryParse(s, out apiError)) {
+                            response.Headers.TryGetValues(CustomHttpHeaders.RTVSBrokerException, out values);
+                            throw new BrokerApiErrorException(apiError, values?.FirstOrDefault());
+                        } else {
+                            throw new ProtocolViolationException("Unknown broker API error");
+                        }
+                    }
+                }
+
+                disposeResponse = false;
+                return response.EnsureSuccessStatusCode();
+            } finally {
+                if (disposeResponse) {
+                    response.Dispose();
+                }
             }
+        }
 
-            IEnumerable<string> values;
-            if(response.Headers.TryGetValues(CustomHttpHeaders.RTVSApiError, out values)) {
-                var s = values.FirstOrDefault();
-                if (s != null) {
-                    BrokerApiError apiError;
-                    if (Enum.TryParse(s, out apiError)) {
-                        response.Headers.TryGetValues(CustomHttpHeaders.RTVSBrokerException, out values);
-                        throw new BrokerApiErrorException(apiError, values?.FirstOrDefault());
-                    } else {
-                        throw new ProtocolViolationException("Unknown broker API error");
+        private async Task<T> RepeatUntilAuthenticatedAsync<T>(Func<Task<T>> action) {
+            while (true) {
+                bool? isValidCredentials = null;
+                try {
+                    _credentialsProvider.UpdateCredentials();
+                    isValidCredentials = true;
+                    return await action();
+                } catch (UnauthorizedAccessException) {
+                    isValidCredentials = false;
+                    continue;
+                } finally {
+                    if (isValidCredentials != null) {
+                        _credentialsProvider.OnCredentialsValidated(isValidCredentials.Value);
                     }
                 }
             }
+        }
 
-            return response.EnsureSuccessStatusCode();
+        private async Task RepeatUntilAuthenticatedAsync(Func<Task> action) {
+            await RepeatUntilAuthenticatedAsync(async () => {
+                await action();
+                return false;
+            });
         }
 
         public async Task<TResponse> HttpGetAsync<TResponse>(Uri uri, CancellationToken cancellationToken = default(CancellationToken)) {
-            var resonse = await HttpClient.GetAsync(uri, cancellationToken);
-            return JsonConvert.DeserializeObject<TResponse>(await resonse.Content.ReadAsStringAsync());
+            using (var response = await RepeatUntilAuthenticatedAsync(async () => EnsureSuccessStatusCode(await HttpClient.GetAsync(uri, cancellationToken)))) {
+                return JsonConvert.DeserializeObject<TResponse>(await response.Content.ReadAsStringAsync());
```

Comment:
No, by the time you get to it, the HTTP response status code and headers have already been checked. This is just reading the body of the response.